### PR TITLE
Improve Story Discoverability on Frontend

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

Sharing stories should render nice previews out of the box. This PR adds Open Graph & Twitter card output to achieve this.

## Relevant Technical Choices

Adds a dedicated `Discovery` class that deals with anything related to `<head>` metadata and such.

## Testing Instructions

---

Addresses #1117

### CODE REVIEW

File: includes/Discovery.php [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+class Discovery {
+	/**
+	 * Initialize discovery functionality.
+	 *
+	 * @return void
+	 */
+	public function init() {
+		add_action(
+			'web_stories_story_head',
+			static function () {
```
Comment (spacedmonkey) [inline includes/Discovery.php:46]: Why is this a static function again?

File: includes/Discovery.php [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+		 * @var WP_Post $post
+		 */
+		$post = get_queried_object();
+
+		if ( $post instanceof WP_Post ) {
+			$metadata = array_merge(
+				$metadata,
+				[
+					'@type'            => 'BlogPosting',
+					'mainEntityOfPage' => get_permalink(),
```
Comment (spacedmonkey) [inline includes/Discovery.php]:
```suggestion
					'mainEntityOfPage' => get_permalink( $post ),
```

File: includes/Discovery.php [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+					'headline'         => get_the_title(),
```
Comment (spacedmonkey) [inline includes/Discovery.php]:
```suggestion
					'headline'         => get_the_title( $post ),
```

File: includes/Discovery.php [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+		<script type="application/ld+json"><?php echo wp_json_encode( $metadata, JSON_UNESCAPED_UNICODE ); ?></script>
+		<?php
+	}
+
+	/**
+	 * Get schema.org metadata for the current query.
+	 *
+	 * @return array $metadata All schema.org metadata for the post.
+	 */
+	protected function get_schemaorg_metadata() {
```
Comment (spacedmonkey) [inline includes/Discovery.php:93]: Seems like this is copied from - https://github.com/ampproject/amp-wp/blob/722f8c9612b2e831217ed9e72ba50caa421f3553/includes/amp-helper-functions.php#L1526-L1594

File: includes/Discovery.php [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+		 *
+		 * @var WP_Post $post
+		 */
+		$post = get_queried_object();
+
+		if ( $post instanceof WP_Post ) {
+			$metadata = array_merge(
+				$metadata,
+				[
+					'@type'            => 'BlogPosting',
```
Comment (spacedmonkey) [inline includes/Discovery.php]: The type was discussed here - https://github.com/Yoast/wordpress-seo/pull/13446#issuecomment-547354523

I think we landed on `article` but it is unclear.

**Comment by spacedmonkey** [comment]
Do you think we have enough filters to intergrate with say yoast or jetpack seo tools?

**Comment by spacedmonkey** [comment]
Related: https://github.com/Yoast/wordpress-seo/pull/13446 https://github.com/google/web-stories-wp/issues/113

File: includes/Discovery.php (hunk 1 above)
Comment (swissspidy) [reply includes/Discovery.php:46]: [Docs](https://www.php.net/manual/en/functions.anonymous.php):

> As of PHP 5.4, anonymous functions may be declared statically. This prevents them from having the current class automatically bound to them. Objects may also not be bound to them at runtime.

Static anonymous functions (closures) have a slight performance benefit over non-static ones. So if we don't need access to `$this` within the function, it is a good practice to declare it as static.

File: includes/Discovery.php (hunk 4 above)
Comment (swissspidy) [reply includes/Discovery.php:93]: Originally, yes. But in this PR I merely moved the function around.

The metadata format will probably change slightly. For example, `BlogPosting` is not entirely accurate for a story.

File: includes/Discovery.php (hunk 5 above)
Comment (swissspidy) [reply includes/Discovery.php]: Yep, we need to change this.

The recommendations are here: 

https://developers.google.com/search/docs/guides/enable-web-stories

It's a bit more complicated, but `Article` as the default makes the most sense.

**Comment by swissspidy** [comment]
> Do you think we have enough filters to intergrate with say yoast or jetpack seo tools?

For now, I think so. But only by collaborating with Yoast & others and putting this out into the wild we'll know for sure. That's what the beta will be useful for.

File: includes/Discovery.php (hunk 5 above)
Comment (swissspidy) [reply includes/Discovery.php]: Changed to `Article`

File: includes/Discovery.php (hunk 4 above)
Comment (spacedmonkey) [reply includes/Discovery.php:93]: Just wonder if it worth having a comment to say that it was taken from there.

File: includes/Discovery.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/Discovery.php:46]: Why is it static at all? Below there are methods accessed on the class. We follow this pattern.

Comment (spacedmonkey) [reply includes/Discovery.php:46]: There should also be test coverage for this.

**Review by spacedmonkey — APPROVED** [review]
There were a couple of things as nitpicks, but nothing that really blocks this being merged.

**Pull request merged**